### ISSUE TITLE
parsePReLU function in caffeParser/opeParsers seems to have a bug (does not work)

### ISSUE THREAD

**Issue by kostinalexey** [NVIDIA/TensorRT#355]
I am struggling to load caffe network with PReLU layer (for example det1 from https://github.com/OAID/FaceDetection/tree/master/models).

The whole function is here:
```
ILayer* parsePReLU(INetworkDefinition& network, const trtcaffe::LayerParameter& msg, CaffeWeightFactory& weightFactory,
                   BlobNameToTensor& tensors)
{
    // Caffe stores the slopes as weights rather than as a tensor, and only supports different slopes
    // per channel
    if (!checkBlobs(msg, 1, 1))
    {
        return nullptr;
    }

    const trtcaffe::PReLUParameter& p = msg.prelu_param();
    bool channelShared = p.has_channel_shared() ? p.channel_shared() : false;
    auto inputDims = tensors[msg.bottom(0)]->getDimensions();
    if (inputDims.nbDims < 2)
    {
        return nullptr;
    }
    int nWeights = channelShared ? 1 : inputDims.d[1]; // Caffe treats second input dimension as channels
    Dims slopesDims{inputDims.nbDims, {1}, {DimensionType::kSPATIAL}};
    slopesDims.d[1] = nWeights;

    Weights w = weightFactory.isInitialized() ? weightFactory(msg.name(), WeightType::kGENERIC) :
                weightFactory.allocateWeights(nWeights, std::uniform_real_distribution<float>(0.F, 1.F));
    auto constLayer = network.addConstant(slopesDims, w);
    return network.addParametricReLU(*tensors[msg.bottom(0)], *constLayer->getOutput(0));
}
```
I can see at least one problem here. For the function to succeed inputDims.nbDims should be >= 2 (In my case it is 3). However slopesDims{inputDims.nbDims, {1}, {DimensionType::kSPATIAL}} initialises only first dimension. The line "slopesDims.d[1] = nWeights;" sets second dimension but not the type. And other dimension(s) are not initialised at all.

This results in failure to create Constant layer later with message " Parameter check failed at: ../builder/Network.cpp::addConstant::770, condition: allDimsGtEq(dimensions, 1)"

**Comment by kostinalexey** [comment]
I found there is a pull request to address the problem. Fix PReLU bug #299.

However it could lead to some other issues.
The line
```
Dims slopesDims{inputDims.nbDims, {nWeights, 1, 1}, {DimensionType::kCHANNEL, DimensionType::kSPATIAL, DimensionType::kSPATIAL}};
```
Is only valid when inputDims.nbDims == 3. However no checks for this done.

**Comment by rmccorm4** [comment]
Hi @kostinalexey, can you comment on the PR with your suggested changes?

Also, if you make these changes locally and re-build the parser, can you parse your model successfully?

**Comment by kostinalexey** [comment]
Hi @rmccorm4, I put my comment on the PR #299. At this stage I can only say that the fix does not deal with all possible values of inputDims.nbDims and this causes segfault in my case. I don't know yet much about guts of TensorRT, but I think the fix should be something like
```
    Dims slopesDims{inputDims.nbDims, {1, 1, 1, 1, 1, 1, 1, 1},
                    {DimensionType::kSPATIAL, DimensionType::kSPATIAL, DimensionType::kSPATIAL, DimensionType::kSPATIAL,
                     DimensionType::kSPATIAL, DimensionType::kSPATIAL, DimensionType::kSPATIAL, DimensionType::kSPATIAL}};
    for(int i = 0; i < inputDims.nbDims; i++) {
        int d{0};
        switch(inputDims.type[i]) {
        case DimensionType::kSPATIAL:
            d = 1; break;
        case DimensionType::kCHANNEL:
            d = channelShared ? 1 : inputDims.d[i]; break;
        case DimensionType::kINDEX:
            d = inputDims.d[i]; break;
        case DimensionType::kSEQUENCE:
            d = 1; break; //I am not sure it is correct. I think caffe does not have this.
        }
        slopesDims.d[i] = d;
        slopesDims.type[i] = inputDims.type[i];
    }
```
However it uses Dims::type member which is labelled as TRT_DEPRECATED. With this code I can at least load the model.

**Comment by rajeevsrao** [comment]
@kostinalexey Thanks - will address this in https://github.com/NVIDIA/TensorRT/pull/299 in the upcoming monthly release

**Comment by kostinalexey** [comment]
Hi Rajeev,

Do you know when this issue will be fixed in the official release of TensorRT?

Regards
Alexey

**Comment by kostinalexey** [comment]
Hi Rajeev,

I have submitted a merge request which fixes the issue #1032.

Regards
Alexey

**Comment by nvpohanh** [comment]
We are deprecating CaffeParser in favor of ONNX. Closing this for now.
